### DISCUSSION TITLE
How / where do I set player properties with PluginPlayer?

### DISCUSSION THREAD

**Discussion by shamseen** [boardgameio/boardgame.io#929] [Q&A] [answered]
Currently the `ctx` player states are initialized from the plugin config but I need certain properties to be populated before I can start the game (each player is a certain color that can access certain parts of the board at a later phase).

1. Where do I update the properties? I'm trying to do it on my board render but `ctx.player` is undefined.
2. _How_ set the individual properties? From the docs, it doesn't seem like `.set()` does that.

As an aside: I'm probably not doing this right, but since each player is a specific color, I want to use the player property `color` to assign the spaces they later have access to.

**Comment by delucis** [comment, marked as answer]
Hey @shamseen the `get`/`set` referred to [in the plugin documentation](https://boardgame.io/documentation/#/plugins?id=pluginplayer) are in moves. For example:

```js
const game = {
  plugins: [
    PluginPlayer({
      setup: () => ({ score: 0 }),
    })
  ],

  moves: {
    updateScore: (G, ctx, newScore) => {
      // Get the stored values for the current player.
      const playerState = ctx.player.get();
      // Update some property.
      playerState.score = newScore;
      // Store the updated state.
      ctx.player.set(playerState);
    }
  }
};
```

Then from your board you would do `moves.updateScore(50)` or whatever to update that bit of player state.

To display player data from your board, you need to access `plugins.player.data`, which will be a plain object that looks something like:

```js
{
  players: {
    0: { score: 50 },
    1: { score: 0 },
  },
}
```

---

As a side note: I personally find this API a bit fiddly. I prefer storing player data manually in the main game state `G`:


```js
const game = {
  setup: (ctx) => {
    let initialState = {
      players: {},
    };
    for (let i = 0; i < ctx.numPlayers; i++) {
      players[i] = { score: 0 };
    }
    return initialState;
  },

  moves: {
    updateScore: (G, ctx, newScore) => {
      // Update some player state:
      G.players[ctx.currentPlayer].score = newScore;
    }
  }
};
```

You’d still call `moves.updateScore(50)` from your board, and would access `G.players` instead of `plugins.player.data.players`.
